_posts/2016-05-05-cargo-pillars.md

```diff
@@ -0,0 +1,489 @@
+time functionality.
+
+First, we'll use `cargo new` to start us out:
+
+```
+$ cargo new datetime
+$ ls
+Cargo.toml src
+$ cat Cargo.toml
+[package]
+name = "datetime"
+version = "0.1.0"
+authors = ["Yehuda Katz <wycats@gmail.com>"]
+
+[dependencies]
+```
+
+We don't want to build the date or time functionality from scratch, so let's
+edit the `Cargo.toml` and add the `time` crate from crates.io:
+
+```
+  [package]
+  name = "datetime"
+  version = "0.1.0"
+  authors = ["Yehuda Katz <wycats@gmail.com>"]
+
+  [dependencies]
++ time = "0.1.35"
+```
+
+Now that we've added the `time` crate, let's see what happens if we ask Cargo to
+build our package:
+
+```
+$ cargo build
+   Compiling winapi v0.2.6
+   Compiling libc v0.2.10
+   Compiling winapi-build v0.1.1
+   Compiling kernel32-sys v0.2.2
+   Compiling time v0.1.35
+   Compiling datetime v0.1.0 (file:///Users/ykatz/Code/datetime)
+```
+
+Whoa! That's a lot of crates. **The biggest part of Cargo's job is to provide
+enough predictability to allow functionality like the `time` crate to be broken
+up into smaller crates that do one thing and do it well**.
+
+Now that we successfully built our crate, what happens if we try to build it again?
+
+```
+$ cargo build
+
+```
+
+Nothing happened at all. Why's that? We can always ask Cargo to give us more
+information through the `--verbose` flag, so let's do that:
+
+```
+$ cargo build --verbose
+       Fresh libc v0.2.10
+       Fresh winapi v0.2.6
+       Fresh winapi-build v0.1.1
+       Fresh kernel32-sys v0.2.2
+       Fresh time v0.1.35
+       Fresh datetime v0.1.0 (file:///Users/ykatz/Code/datetime)
+```
+
+Cargo isn't bothering to recompile packages that it knows are "fresh". If you've
+ever had to configure a `Makefile` before, this should be a breath of fresh air
+(no pun intended).
+
+But how does Cargo know that everything is fresh? When Cargo builds a crate, it
+emits a file called `Cargo.lock` that contains the precise versions of all of
+its resolved dependencies:
+
+```
+[root]
+name = "datetime"
+version = "0.1.0"
+dependencies = [
+ "libc 0.2.10 (registry+https://github.com/rust-lang/crates.io-index)",
+ "time 0.1.35 (registry+https://github.com/rust-lang/crates.io-index)",
+]
+
+[[package]]
+name = "kernel32-sys"
+version = "0.2.2"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+dependencies = [
+ "winapi 0.2.6 (registry+https://github.com/rust-lang/crates.io-index)",
+ "winapi-build 0.1.1 (registry+https://github.com/rust-lang/crates.io-index)",
+]
+
+...
+```
+
+The `Cargo.lock` contains a serialized version of the entire **resolved**
+dependency graph, including precise versions of all of the source code included
+in the build. In the case of a package from crates.io, Cargo stores the name and
+version of the dependency. This is enough information to uniquely identify
+source code from crates.io, because the registry is **append only** (no changes
```

Comment:
Perhaps crates.io could be linkified to crates.io and we could mention the index and linkify that?